build: bump motsu v0.4.0 (#39)
OpenZeppelin/stylus-test-helpers: crates/motsu/README.md

## Usage

This crate enables unit-testing for Stylus contracts. It abstracts away the
machinery necessary for writing tests behind a `#[motsu::test]` procedural macro.

Annotate tests with `#[motsu::test]` instead of `#[test]`
to get access to VM affordances:

>  ```rust
>  #[cfg(test)]

Reviewer: ```suggestion
```rust
#[cfg(test)]
```

mod tests {

>     use openzeppelin_stylus::token::erc20::Erc20;

Reviewer: ```suggestion
    use openzeppelin_stylus::token::erc20::{Erc20, IErc20};
```
`balance_of` is only available with the `IErc20` trait imported

 ```rust
use motsu::prelude::*;

#[motsu::test]
fn call_three_proxies(
    proxy1: Contract<Proxy>,
    proxy2: Contract<Proxy>,
    proxy3: Contract<Proxy>,
    alice: Account,
) {
    let one = uint!(1_U256);
    let ten = uint!(10_U256);

    // Set up a chain of three proxies.
    // With the given call chain: proxy1 -> proxy2 -> proxy3.
    proxy1.sender(alice).init(proxy2.address());
    proxy2.sender(alice).init(proxy3.address());
    proxy3.sender(alice).init(Address::ZERO);

    // Call the first proxy.
    let result = proxy1.sender(alice).call_proxy(ten);

    // The value is incremented by 1 for each proxy.
    assert_eq!(result, ten + one + one + one);
}
 ```

Annotating a test function that accepts no parameters will make
`#[motsu::test]` behave the same as `#[test]`.

 ```rust,ignore
 #[cfg(test)]
mod tests {
    #[motsu::test] // Equivalent to #[test]
    fn test_fn() {
        ...
    }
}
 ```

NOTE!!!
We require a contract to implement unsafe trait
`stylus_sdk::prelude::TopLevelStorage`, for a contract to be used in tests.
Typically, all contracts marked with `#[entrypoint]` will have this trait
automatically derived.
Otherwise, you should do it by

> yourself:

Reviewer: Good note! Though the `NOTE!!!` part is a bit unprofessional, like we're yelling.

I'd suggest any of the following:
- `**Important:** <rest of important note>`
- `NOTE: <rest of note>`
- `> NOTE: <rest of note>` - make it a quote to make it "pop out"

Author: Let's add "**Important:**`

> We require a contract to implement unsafe trait
> `stylus_sdk::prelude::TopLevelStorage`, for a contract to be used in tests.
> Typically, all contracts marked with `#[entrypoint]` will have this trait
> automatically derived.
> Otherwise, you should do it by
> yourself:

Reviewer: ```suggestion
To use a contract in tests, you must ensure it implements the unsafe
trait `stylus_sdk::prelude::TopLevelStorage`. While this trait is automatically
derived for contracts marked with `#[entrypoint]`, you'll need to implement
it manually for any contract without this attribute. This trait is essential
for managing contract storage during testing.
```
- explicit context about the purpose
- clearer wording

Author: Have changed it in `lib.rs` also, but without last sentence.

## Usage

This crate enables unit-testing for Stylus contracts. It abstracts away the

> machinery necessary for writing tests behind a

Reviewer: ```suggestion
logic necessary for writing tests behind a
```
We have heavy "machinery" :smile:

Author: That is a more `fancy` language and I'm fine with it. It wasn't me who added it before:D

[`#[motsu::test]`][test_attribute] procedural macro.

Annotate tests with [`#[motsu::test]`][test_attribute] instead of `#[test]`
to get access to VM affordances:

 ```rust
 #[cfg(test)]
mod tests {
    use motsu::prelude::*;
    use stylus_sdk::alloy_primitives::{Address, U256};

    #[motsu::test]
    fn reads_balance(

>         contract: Contract<Erc20>,

Reviewer: Would be helpful to have a "dumb" Erc20 contract definition above this test, just to have a simple working example of how Motsu could be used

Reviewer 2: Love free marketing! Let's do it 🚀

        alice: Account,
    ) {
        // Access storage.
        let balance = contract.sender(alice).balance_of(Address::ZERO);
        assert_eq!(balance, U256::ZERO);
    }
}
 ```

Function `Contract::sender()` is necessary to trigger call
to a contract, and should accept an `Account` or `Address` as an argument.

Alternatively `Contract::sender_and_value()` can be used to
pass additional value to the contract:

 ```rust

> use motsu::prelude::*;
> use stylus_sdk::alloy_primitives::{Address, U256, ruint::uint};
> 
> #[motsu_proc::test]

Reviewer: ```suggestion
#[motsu::test]
```
1. I wouldn't show imports in every example. Devs using Stylus SDK would already be familiar with these types.
2. Forgot to apply the [previous suggestion](https://github.com/OpenZeppelin/stylus-test-helpers/pull/39/files#r1945255670).

Author: in `README.md` probably make sense

> fn pay_three_proxies(proxy: Contract<Proxy>, alice: Account) {

Reviewer: Where did this `Proxy` contract come from? What does it's `pay_proxy()` function do? This will confuse docs readers.
Either implement a simple `Proxy` contract or mention what it does in the description above the code example.

Reviewer: https://github.com/OpenZeppelin/stylus-test-helpers/pull/39#discussion_r1946309375

Author: Just how `sender_with_value(..)` should be invoked

    let one = uint!(1_U256);
    let ten = uint!(10_U256);

    // Initialize the proxy contract.
    proxy.sender(alice).init(Address::ZERO);

    // Fund alice.

>     alice.fund(ten);

Reviewer: Where did `fund` come from? We should mention that mocking functions exist prior to using them in examples.

Author: documented it

    // Call the contract.
    proxy.sender_and_value(alice, one).pay_proxy();

    // Assert that alice lost one wei and the proxy gained one wei.
    assert_eq!(alice.balance(), ten - one);
    assert_eq!(proxy.balance(), one);
}
 ```

Multiple external calls are supported in Motsu. Assuming `Proxy` is a
contract that exposes (`#[public]`) function `call_proxy`. Where it adds `one`
to the passed argument and calls next `Proxy` contract at the address
provided during initialization. The following test case can emulate a call

> chain of three `Proxy` contracts:

Reviewer: This should be placed before the above code example (where the contract is first used)

Reviewer: https://github.com/OpenZeppelin/stylus-test-helpers/pull/39#discussion_r1946295575

Author: I don't think so. It describes example below quite well.

Author: Example above is mostly about calling contract with value

Reviewer: Yes, but the example is introducing too many previously unmentioned items:
- `Proxy`
- `Proxy::init`
- `fund` (related [comment](https://github.com/OpenZeppelin/stylus-test-helpers/pull/39/files#r1946308354))
- `Proxy::pay_proxy`

How is a dev to know which of these is part of motsu and which is just an example?

Author: Let's revisit better examples, when change our test cases in `opnezeppelin-stylus` crate

    // Call the first proxy.
    let result = proxy1.sender(alice).call_proxy(ten);

    // The value is incremented by 1 for each proxy.
    assert_eq!(result, ten + one + one + one);
}
 ```

Annotating a test function that accepts no parameters will make
[`#[motsu::test]`][test_attribute] behave the same as `#[test]`.

 ```rust,ignore
 #[cfg(test)]
mod tests {
    #[motsu::test] // Equivalent to #[test]
    fn test_fn() {
        ...
    }
}
 ```

NOTE!!!
We require a contract to implement unsafe trait
`stylus_sdk::prelude::TopLevelStorage`, for a contract to be used in tests.
Typically, all contracts marked with `#[entrypoint]` will have this trait
automatically derived.
Otherwise, you should do it by
yourself:

 ```rust
 use stylus_sdk::{
    storage::{StorageMap, StorageU256, StorageAddress},
    prelude::*,
    alloy_primitives::Address,
};

// Entry point is not implemented, so we should implement `TopLevelStorage` ourselves.
// #[entrypoint]
#[storage]
pub struct Erc20 {
    balances: StorageMap<Address, StorageU256>,
    allowances: StorageMap<Address, StorageMap<Address, StorageU256>>,
    total_supply: StorageU256,
}

unsafe impl TopLevelStorage for Erc20 {}
 ```

> 
> NOTE!!!
> For `motsu` to work correctly, `stylus-sdk` should **not** have
> default `hostio-caching` feature enabled.

Reviewer: ```suggestion
```
This was [already mentioned](https://github.com/OpenZeppelin/stylus-test-helpers/pull/39/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R23-R24)

> 
> Following features: event log assertions, storage reset after erroneous transaction,
> `proptest` affordances for `Contract` and `Account`
> are not supported, and planed to be added soon.

Reviewer: ```suggestion
```
If devs are interested in our roadmap, they should check our issues page.

We can mention the features that _are_ supported though.

Author: For people not to search much. But the downside we can forget to remove these ones

## Usage

This crate enables unit-testing for Stylus contracts. It abstracts away the
machinery necessary for writing tests behind a
[`#[motsu::test]`][test_attribute] procedural macro.

Annotate tests with [`#[motsu::test]`][test_attribute] instead of `#[test]`
to get access to VM affordances:

 ```rust
 #[cfg(test)]
mod tests {
    use motsu::prelude::*;
    use stylus_sdk::alloy_primitives::{Address, U256};

    #[motsu::test]
    fn reads_balance(
        contract: Contract<Erc20>,
        alice: Account,
    ) {
        let balance = contract.sender(alice).balance_of(Address::ZERO); // Access storage.
        assert_eq!(balance, U256::ZERO);
    }
}
 ```

Function [`crate::prelude::Contract::sender`] is necessary to trigger call
to contract, and should accept an account or address as an argument.

Alternatively [`crate::prelude::Contract::sender_and_value`] can be used to
pass additional value to the contract:

 ```rust
  use motsu::prelude::*;
use stylus_sdk::alloy_primitives::{Address, U256, ruint::uint};

> #[motsu_proc::test]

Reviewer 2: ```suggestion
#[motsu::test]
```

Reviewer: Unresolving, as the change wasn't applied nor addressed

fn pay_three_proxies(proxy: Contract<Proxy>, alice: Account) {
    let one = uint!(1_U256);
    let ten = uint!(10_U256);

    // Initialize the proxy contract.
    proxy.sender(alice).init(Address::ZERO);

    // Fund alice.
    alice.fund(ten);

    // Call the contract.
    proxy.sender_and_value(alice, one).pay_proxy();

    // Assert that alice lost one wei and the proxy gained one wei.
    assert_eq!(alice.balance(), ten - one);
    assert_eq!(proxy.balance(), one);
}
 ```

Multiple external calls are supported in Motsu. Assuming `Proxy` is a
contract that exposes `#[public]` function `call_proxy`. Where it adds `one`
to the passed argument and calls next `Proxy` contract at the address
provided during initialization. The following test case can emulate a call
chain of three `Proxy` contracts:

 ```rust

>   use motsu::prelude::*;

Reviewer 2: ```suggestion
use motsu::prelude::*;
```

use stylus_sdk::alloy_primitives::{Address, U256, ruint::uint};

> #[motsu_proc::test]

Reviewer 2: ```suggestion
#[motsu::test]
```

Reviewer: Unresolving as the change was neither applied nor addressed
